core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -562,6 +563,8 @@ class SparkContext(config: SparkConf) extends Logging {
       }
     _cleaner.foreach(_.start())
 
+    _stopHooks = new SparkShutdownHookManager()
```

Comment:
there's already a shutdown hook to call sc.stop() - perhaps just add the clean up in stop()
https://github.com/aray/spark/blob/005472ed10fad3d1bc8feff12fc55c5682724a0e/core/src/main/scala/org/apache/spark/SparkContext.scala#L584

Reply:
The queries also need to be gracefully stopped if someone calls `sc.stop()` without shutting down the JVM.

Reply:
I think that's my point.
if sc.stop() calls streams.stopAllQueries()
then if someone calls sc.stop() it works
if someone shutdown the JVM, sc.stop() is already getting called.

Reply:
@felixcheung I don't get the objection. This is not adding a new shutdown hook, just reusing the class to add "stop hooks" that are called by `sc.stop()`. Furthermore this is needed because `sc.stop()` can not just call `stopAllQueries` unassisted without creating a circular package dependency.